packages/ui/src/providers/RouteCache/index.tsx

```diff
@@ -19,13 +19,29 @@ export const RouteCache: React.FC<{ cachingEnabled?: boolean; children: React.Re
 }) => {
   const pathname = usePathname()
   const router = useRouter()
+  const clearAfterPathnameChange = useRef(false)
 
   const clearRouteCache = useCallback(() => {
     router.refresh()
   }, [router])
 
   useEffect(() => {
-    if (cachingEnabled) {
+    const handlePopState = () => {
+      // Calling `router.refresh()` directly here doesn't work. Probably fires too early.
+      // Instead, need to set a flag that we can check on the next pathname change.
+      clearAfterPathnameChange.current = true
+    }
+
+    window.addEventListener('popstate', handlePopState)
+
+    return () => {
+      window.removeEventListener('popstate', handlePopState)
+    }
+  }, [router])
+
+  useEffect(() => {
+    if (cachingEnabled || clearAfterPathnameChange.current) {
+      clearAfterPathnameChange.current = false
       clearRouteCache()
     }
   }, [pathname, clearRouteCache, cachingEnabled])
```

Comment:
I think this is a good candidate for `useEffectEvent` to make sure this runs more predictably and only if pathname changes